### PR TITLE
Fix revocation of leases when num_uses goes to 0

### LINKED ISSUE #2189: accessor leak
Hi,

We are using a lot of short lived tokens that are generated with the response wrapping.
The problem is that Vault never deletes the accessor of the temp tokens and the number of accessors keeps growing in consul.
With time we got to around 300k accessors not accounted for. (number of accessors - number of tokens). Vault is at the brink of not being able to take over the active role when it has to read all those accessors into memory.

We see it in vault 0.6.1 and 0.6.3

### How to reproduce
* Generate wrapped token (its the same when using `sys/wrapping/wrap`)
  ```
  $ vault token-create -format=json -policy="default" -wrap-ttl 300 -orphan
  {
  	"request_id": "",
  	"lease_id": "",
  	"lease_duration": 0,
  	"renewable": false,
  	"data": null,
  	"warnings": null,
  	"wrap_info": {
  		"token": "3dfe8a33-3151-e865-dbb4-7dd32bfd2b0d",
  		"ttl": 300,
  		"creation_time": "2016-12-15T15:55:11.754551887Z",
  		"wrapped_accessor": "96f88426-a083-7e31-debd-e46780184220"
  	}
  }
  ```
  
  In consul I can see the perm and temp tokens being created with the associated accessors. The accessor list grows by 2 (`vault list auth/token/accessors | wc -l`).
  
  ```
  PUT /v1/kv/vault/sys/token/accessor/f3302d70275c55c03e4216c97c0acd3f64ed80ec 
  PUT /v1/kv/vault/sys/token/id/5bd71de229098b14af879df51448e05ecafaad15 
  PUT /v1/kv/vault/sys/token/accessor/95c17930b42fe0902e3ea4f5a750bf677446ec7a 
  PUT /v1/kv/vault/sys/token/id/bfdd2105bbbace3a68e4ccb50b4d7d39d1959efa
  ```
  The temp accessor when looked-up gives: `"policies": [ "response-wrapping" ]` as expected.

* unwrap the perm token
  ```
  $ vault unwrap -format=json 3dfe8a33-3151-e865-dbb4-7dd32bfd2b0d
  {
  	"request_id": "6b99824d-b9cd-9eba-d80a-7a897439579b",
  	"lease_id": "",
  	"lease_duration": 0,
  	"renewable": false,
  	"data": null,
  	"warnings": null,
  	"auth": {
  		"client_token": "056eaed7-b241-d70b-b1c8-f72656e645ed",
  		"accessor": "96f88426-a083-7e31-debd-e46780184220",
  		"policies": [
  			"default"
  		],
  		"metadata": null,
  		"lease_duration": 86400,
  		"renewable": true
  	}
  }
  ```
  
  In consul I can see just the temp token being deleted while its accessor is left behind.

  ```
  DELETE /v1/kv/vault/sys/token/id/bfdd2105bbbace3a68e4ccb50b4d7d39d1959efa 
  ```

  The temp accessor when looked-up gives: `1 error occurred:\n\n* permission denied`. Its there but not accessible or revokable.

### PR DESCRIPTION
**Pull request by jefferai**
Also fixes #2189

### CODE REVIEW

File: vault/token_store.go [hunk 1]
```diff
@@ -869,46 +868,112 @@ func (ts *TokenStore) Revoke(id string) error {
...
+		lock.Unlock()
 		return err
 	}
 
-	// Nuke the primary key first
-	path := lookupPrefix + saltedId
-	if ts.view.Delete(path); err != nil {
-		return fmt.Errorf("failed to delete entry: %v", err)
+	if entry == nil {
+		lock.Unlock()
```
Comment (vishalnayak) [inline vault/token_store.go]: Same question here, is there value in deleting a non-existent token entry?

File: vault/token_store.go [hunk 2]
```diff
@@ -783,11 +805,13 @@ func (ts *TokenStore) Lookup(id string) (*TokenEntry, error) {
 	lock.RLock()
 	defer lock.RUnlock()
 
-	return ts.lookupSalted(ts.SaltID(id))
+	return ts.lookupSalted(ts.SaltID(id), false)
 }
 
-// lookupSlated is used to find a token given its salted ID
-func (ts *TokenStore) lookupSalted(saltedId string) (*TokenEntry, error) {
+// lookupSalted is used to find a token given its salted ID. If tainted is
```
Comment (vishalnayak) [inline vault/token_store.go]: Can we update this comment?

File: vault/token_store.go [hunk 3]
```diff
@@ -869,46 +893,109 @@ func (ts *TokenStore) Revoke(id string) error {
 
 // revokeSalted is used to invalidate a given salted token,
 // any child tokens will be orphaned.
-func (ts *TokenStore) revokeSalted(saltedId string) error {
+func (ts *TokenStore) revokeSalted(saltedId string) (ret error) {
```
Comment (vishalnayak) [inline vault/token_store.go]: [Optional] s/ret/somethingElse

File: vault/token_store.go [hunk 4]
```diff
@@ -869,46 +893,109 @@ func (ts *TokenStore) Revoke(id string) error {
...
+	}()
+
+	// Destroy the token's cubby. This should go first as it's a
+	// security-sensitive item.
+	err = ts.cubbyholeDestroyer(ts, saltedId)
+	if err != nil {
+		return err
+	}
+
+	// Revoke all secrets under this token. This should go first as it's a
```
Comment (vishalnayak) [inline vault/token_store.go]: Can we update the comment? This is not the first anymore.

File: vault/token_store.go [hunk 5]
```diff
@@ -1637,8 +1724,13 @@ func (ts *TokenStore) handleLookup(
...
 	}
 
+	lock := ts.getTokenLock(id)
+	lock.RLock()
+	defer lock.RUnlock()
+
 	// Lookup the token
-	out, err := ts.Lookup(id)
+	saltedId := ts.SaltID(id)
+	out, err := ts.lookupSalted(saltedId, true)
```
Comment (vishalnayak) [inline vault/token_store.go]: I assume we moved from Lookup to lookupSalted to set `tainted` to `true`. I am curious to know why we need to set tainted while doing a `handleLookup`.

File: vault/token_store.go (hunk 4 above)
Comment (jefferai) [inline vault/token_store.go]: I kept that because those two items are both first relative to the rest.

File: vault/token_store.go (hunk 5 above)
Comment (jefferai) [inline vault/token_store.go]: So that a lookup of a token in a revoking state still returns the token information, for debugging/tracking purposes.

**Pull request merged**